src/arithmetic/arithmetic_expression.c

```diff
@@ -477,9 +479,10 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc) {
 
     assert(start >=0 && length >=0);
     
-    substring = calloc(length+1, sizeof(char));
+    substring = malloc(length+1 * sizeof(char));
     memcpy(substring, original + start, length);
```

Comment:
Need to make sure start is within bounds of original's length.
Same goes for start+length this must overreach.

Reply:
The overflow risks have already been validated by a few assert statements a bit higher up in this function. I did rework the logic a bit just now to fix some negative checks against unsigned types, and added one additional test for this function.